### PR TITLE
feat: add useProjectApi for extensions

### PR DESCRIPTION
**Pull request by a-b-r-o-w-n**
## Description

Adds `useProjectApi` hook for ui extensions.

## Task Item

refs #4357

### CODE REVIEW

File: Composer/packages/client/src/components/PluginHost/PluginHost.tsx [hunk 1]
```diff
@@ -39,38 +45,52 @@ function injectScript(doc: Document, id: string, src: string, async: boolean, on
...
+
+      attachPluginAPI(iframeWindow, pluginType, shell);
+
+      //load the bundle for the specified plugin
+      const pluginScriptId = `plugin-${pluginType}-${pluginName}`;
+      const bundleUri = `/api/extensions/${pluginName}/${bundleId}`;
+      // If plugin bundles end up being too large and block the client thread due to the load, enable the async flag on this call
+      injectScript(iframeDocument, pluginScriptId, bundleUri, false);
+    }
+  }, [isLoaded]);
```
[the remark this answers is not in the document]
Comment (a-b-r-o-w-n) [reply Composer/packages/client/src/components/PluginHost/PluginHost.tsx:64]: I'll update.

File: Composer/packages/extension-client/src/hooks/useProjectApi.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+export function useProjectApi(): ProjectContext & ProjectContextApi {
+  const shell = useStore();
+
+  const projectContext = useMemo(() => {
+    const ctx = pick(shell, PROJECT_KEYS);
+    return {
+      ...ctx.api,
+      ...ctx.data,
+    } as ProjectContext & ProjectContextApi;
+  }, [pick(shell, PROJECT_KEYS)]);
```
[the remark this answers is not in the document]
Comment (a-b-r-o-w-n) [reply Composer/packages/extension-client/src/hooks/useProjectApi.ts:48]: I want to scope it down to only the properties this context cares about.

File: Composer/packages/extension-client/src/store.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+class ExtensionStore implements Store {
+  private state: State = {} as Shell;
+  private listeners: Dispatch<Partial<State>>[] = [];
+
+  constructor(initialState = {}) {
+    const windowState = window[ComposerGlobalName]?.[STORAGE_KEY] ?? {};
+
+    if (window[ComposerGlobalName]?.[STORAGE_KEY]) {
+      this.setState(merge({}, windowState, initialState));
```
[the remark this answers is not in the document]
Comment (a-b-r-o-w-n) [reply Composer/packages/extension-client/src/store.ts:27]: I may not even need the window state anymore. Let me check.

File: Composer/packages/client/src/components/PluginHost/PluginHost.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (a-b-r-o-w-n) [reply Composer/packages/client/src/components/PluginHost/PluginHost.tsx:64]: I don't think I need to do this here. The `isLoaded` flag is set inside of the below `useEffect` which has a dependency on those props.

[the remark this answers is not in the document]
Comment (hatpick) [reply Composer/packages/client/src/components/PluginHost/PluginHost.tsx:64]: Given that the effects run in order, maybe move the one that updates the isLoaded above this one so it doesn't cause this effect two run more than once

**Review by hatpick — APPROVED** [review]

**Pull request merged**